### ISSUE TITLE
Make OpenSim crash when it fails to load a component from a .osim file

### ISSUE THREAD

**Issue by adamkewley** [opensim-org/opensim-core#2911]
Hi,

**Can we *please* make it an error, rather than a warning, to load an OpenSim file that contains components that are not available in OpenSim**?

This issue costed me a fair bit of time to diagnose today. It isn't automatically obvious when it happens and it can happen for reasons unrelated to OpenSim (specifically: the linker optimizing away libraries when no symbols are used from them).

# Core of the Issue

Pretend I'm opening a modified arm26 ([here](https://raw.githubusercontent.com/opensim-org/opensim-models/master/Models/Arm26/arm26.osim)) which contains `BsMuscle`: an extra, non-existent, muscle component:

```xml
<!--arm26.osim:1045 -->
<ForceSet>
    <objects>
        <BsMuscle></BsMuscle> 

    <!-- rest of the file, etc. -->
```

I'd expect that loading this in OpenSim causes an error, because the model is not "complete" without this muscle. Instead, though, it produces a warning in the application's output:

```
[info] Updating Model file from 20303 to latest format...
Object::newInstanceOfType(): object type 'BsMuscle' is not a registered Object!
Object type BsMuscle not recognized
[info] Loaded model arm26 from file arm26.osim
# .... a lot of other warnings about things like VTP files missing, etc....
```

Which isn't easy to notice because it gets buried in all the other warnings (plus, we're diagnosing problems via the OpenSim visualizer, which doesn't show this issue). 

What this means is that the API thinks the model loading was successful (albeit, with warnings). Downstream code *may* then fail at runtime if the runtime code is reliant on those components being loaded. However, downstream code may also run to completion and fail indirectly, which was what ate our time.

I believe that the *vast majority* of OpenSim users will not notice this, or understand the implications based on that warning message alone. This behaviour means that (e.g.) typical users can load a file containing plugin-based components (e.g. `ScapulothoracicJoint`) and end up with a model that loads with no error, but is functionally incorrect.

# Breakdown of our particular issue

Here is a summary of what happened today because failing to load model components was treated as a warning, rather than an error. 

For context, it took 2-3 hours for two OpenSim developers to finally understand the entire problem and fix it properly. The time taken isn't entirely due to this issue, but due to how this issue interacts with how Ubuntu 18, specifically, will not link a library unless it uses symbols directly from that library. 

This breakdown is my braindump. Other developers may have resolved the issue quicker/slower, but I believe a typical junior/intermediate C++ programmer who hasn't used OpenSim much would spend at least an equivalent amount of time fixing an issue like this.

1. We loaded a standard OpenSim file via the C++ API. Specifically, Rajagopal 2015 ([here](https://github.com/opensim-org/opensim-models/blob/master/Models/RajagopalModel/Rajagopal2015.osim)). This worked fine and produced the usual gamut of warnings (plus the above one, although buried)
2. The code we're running iterates over `OpenSim::Muscle`s in  the model using `Component::getComponentList<T>` to aggregate some useful information. 
3. The code ran to completion with no errors, but the output file was empty
4. (debugging begins)
5. The output was empty because no muscles were iterated over
6. We made sure our code was using the OpenSim API correctly (good thing to check: it's experimental code)
7. We made sure the Rajagopal model contained muscles (it did)
8. We made sure we were using a stable OpenSim version (we were)
9. We visualized the model via the same API. The visualized model contained no muscles (uh)
10. We tried another model file that we know *for a fact* contains muscles because we see the muscles in OpenSimGUI. This model contained no muscles either (what?)
11. We tried stripping our implementation down to a bare-minimum implementation that couldn't feasibly contains bugs. The resulting model still contains no muscles (tf?)
12. We start digging through logs and notice that `Object::newInstanceOfType(): object type 'MillardEquilibriumMuscle2012' is not a registered Object!` is buried in there as a warning (wait, the OpenSim library we linked doesn't support muscles!?)
13. We figured that `MillardEquilibriumMuscle2012` is actually in a separate library from the core OpenSim one (`libActuators.so`). 
    - Sidenote: why is this a separate library, rather than just shipping everything in a single OpenSim library? Are people running OpenSim on memory-constrained devices that can't afford to `mmap` a fatter library?)
14. We diagnosed our (downstream) build. It wasn't linking `osimActuators`, but compiles because it doesn't directly use symbols in `osimActuators`.
15. Added `osimActuators` to the `target_link_libraries` of our target (ah, problem solved!)
16. This version works on my Debian9 machine, but doesn't work on the other dev's Ubuntu18 one (wtf!?)
17. The other dev's executable does not link to `osimActuators`, despite being told to explicitly do so (!?!?)
18. We spend ages trying different linker flags, different library paths, rebuilding from scratch, clean-building, etc. (cmon, surely CMake just hasn't noticed our `osimActuators` addition...)
18. This turns out to be an artefact of Ubuntu18, which links libraries with `--as-needed`. That flag means that dynamic libraries are only linked to an executable if the executable uses symbols in the library.
19. That linker behavior completely screws with OpenSim's design, which leans on static initializers
20. Final solution is to compile with all the necessary libraries and manually provide `-Wl,--no-as-needed` to the linker to ensure that all opensim libraries are linked, even if not directly used.

This may not sound like a huge amount of work overall (check some logs, check some builds) but enhance each of these steps with multiple rounds of "cmon, `ldd` the binary now", "lets see if CMake has cached something", "are you sure you're using a clean version of OpenSim" and it can easily drag on.

---

So, in conclusion, can we please just make loading a non-existent `Component` a hard-error, rather than a warning? If developers/advanced OpenSim users need the warning fallback (e.g. because they rely on designing models with not-yet-developed components) then I'd recommend adding the behavior via a flag (e.g. `OPENSIM_NO_MISSING_COMPONENT_ERROR`?)

**Comment by chrisdembia** [comment]
There's an issue and a WIP PR for this :) 

- https://github.com/opensim-org/opensim-core/issues/686
- https://github.com/opensim-org/opensim-core/pull/803

The issue, if I recall correctly, is that quite a few tests depend on the permissive/silent behavior of our XML parser. Therefore, we should make the strict parsing an option (true by default, ideally).

**Comment by adamkewley** [comment]
Oh god. This was known about in 2015, worked on in 2016, and has now slipped into the void. Since then, it's probably caused a few headaches for other (less vocal) devs, and costed me steps 1-13 above.

I *might* be able to kick that PR, or a fresh PR, into shape by bootstrapping the functionality incrementally (e.g. by making it an explicit API opt-in, followed by an env var opt-in, followed by default opt-in with an env-var opt-out). I imagine the reason this has fell by the wayside is because rolling it out as an explicit-opt-(in/out) manages to annoy everyone:

- If it's default opt-in it breaks tests, so the resulting PR becomes massive and downstreamers suddenly have broken systems
- If it's default opt-out (now) downstreamers like me waste loads of time debugging a related issue
- If it's explicit opt-in then noone will use it and it'll just be weaponized as a "ha ha ha you didn't call this function" retort in later issues
- If it's explicit opt-out then any downstreamers will complain that they need to recompile to make their codebases work again
- If it's envvar opt-in then noone will use it
- If it's envvar opt-out (where we want to be) then any downstreamers will complain, but at least have a recompilation-free solution

So any (small PR) solution will have to incrementally fix the issue over several PRs, followed by making it default-opt-in in a later release (with the option to override that /w a flag/environment variable - and that option being explicitly mentioned in any runtime error message).

**Comment by aymanhab** [comment]
Thanks for the investigation @adamkewley and sorry it took us years to push the PR through. I have a better understanding of the context so I'll open a PR to take care of this and will keep you in the loop.

**Comment by adamkewley** [comment]
This has now been merged, so closing

**Comment by tkuchida** [comment]
Just throwing an Exception may not be most convenient for GUI users. Perhaps the GUI can catch this and show a warning, with options "Cancel" and "Proceed anyway"? See Forum post here: https://simtk.org/plugins/phpBB/viewtopicPhpbb.php?f=91&t=13466
